rfcs/0051-limit-zkapp-cmds-per-block.md

```diff
@@ -0,0 +1,156 @@
+the aforementioned issue until a proper solution can be devised and
+implemented. The root cause of the issue is that proofs contained
+within these commands are stored in the scan state and tend to occupy
+a lot of space. Fixing these storage issues won't affect the
+protocol, so ideally we want a workaround that doesn't affect the
+protocol either, so that at the convenient time we can turn it off
+without making a fork.
+
+## Detailed design
+
+Since the solution should not affect the protocol, it should be
+implemented at the mempool/block producer boundary. In the mempool
+there is `transactions` function, which returns a sequence of
+transactions from the mempool in the order of decreasing transaction
+fees. The block producer module just takes a number of transactions
+from that sequence and includes them. It is relatively easy to limit
+the number of zkApp commands returned by this functions.  ZkApp
+commands above a certain limit would simply be discarded (from the
+sequence, but not from the mempool).
+
+The exact number of zkApps allowed in each block should be set
+dynamically, so that we can adjust it without redeploying nodes.
+The easiest way to control it is through a command line argument
+to the node. Thus changing the setting is as easy as restarting
+the node.
+
+The setting can be stored in the mempool configuration and
+initialized when the mempool is being created at startup.
+It will likely involve some plumbing to transport the setting
+from the command line arguments to the function which initializes
+the mempool, but other than that the solution is quite
+straightforward to implement.
+
+## Drawbacks
+
+Any non-protocol-level solution to this issue has a drawback that a
+malicious node operator could modify their node to turn off the
+safeguard. However, because the safeguard only affects block
+production, it doesn't really matter unless the malicious agent is
+going to produce blocks. If so, their chance of conducting a
+successful DoS attack against the network is proportional to their
+stake, but their incentive to do so is **inversely** proportional
+to their stake, which means the more capable one is to conduct the
+attack, the more they are going to lose in case of success.
+
+With the safeguard turned on, if the zkApps are coming in faster than
+they can be processed, they will stack up in nodes' mempools.
+Mempools **will** eventually overflow, which means that either some of
+these zkApp commands or some regular user commands will start to
+drop. This will likely inflate transaction fees as users will attempt
+to get their transactions into increasingly crowded mempools. Also a
+lot of transactions will be lost in the process due to mempool
+overflow.
+
+The default would be no limit, as that's the easiest to express.
+Alternatively, we can provide another default, but that would require
+providing an option to disable the limit altogether, complicating the
+interface.
```

Comment:
You could have the flag takes options such as `No-limit | Zkapps-16 | Zkapps-24 etc` and have a default value of what's recommended. This is nice because then you have a small set of configurations for which memory requirements can be determined/prescribed.

Reply:
I disagree. The number of zkApp commands in the scan state (which is the root cause of the issue) is affected not only by block production, but also by SNARK production and by which SNARKs a block producer decides to purchase. In fact, even with the limit it still can happen that the scan state ends up filled with zkApp commands, especially if their SNARKs are significantly more expensive. Therefore the limit alone is not sufficient to predict or model the memory requirements of the scan state. For this reason I think we want to have as fine-grained control over the limit as posible. I heard that in addition to the command-line flag we also want to be able to change the limit at runtime (for debugging purposes). If that's the case, it's another argument to retain as precise control as possible.

I will make it so that this option will accept either an integer or a string `none` or `null`, which will introduce no limit. Then we can choose whatever default we like.

Reply:
Okay then. What set of limits do we want?